sound/soc/sof/trace.c

```diff
@@ -13,6 +13,216 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+#define TRACE_FILTER_MAX_IPC_ELEMENTS_PER_ENTRY 4
+
+static int trace_filter_append_elem(int32_t key, int32_t value,
+				    struct sof_ipc_trace_filter_elem *elem_list,
+				    size_t capacity, size_t *counter)
+{
+	if (*counter >= capacity)
+		return -ENOMEM;
+
+	elem_list[*counter].key = key;
+	elem_list[*counter].value = value;
+	++*counter;
+
+	return 0;
+}
+
+static int trace_filter_parse_entry(struct snd_sof_dev *sdev, const char *line,
+				    struct sof_ipc_trace_filter_elem *elem,
+				    size_t capacity, size_t *counter)
+{
+	int len = strlen(line);
+	size_t cnt = *counter;
+	int log_level;
+	int uuid_id;
+	int pipe_id;
+	int comp_id;
+	int read;
+	int ret;
+
+	ret = sscanf(line, " %d %x %d %d %n", &log_level, &uuid_id, &pipe_id, &comp_id, &read);
+	if (ret != 4 || read != len) {
+		dev_err(sdev->dev, "error: invalid trace filter entry '%s'\n",
+			line);
+		return -EINVAL;
+	}
+
+	if (uuid_id > 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_UUID,
+					       uuid_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+	if (pipe_id >= 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_PIPE,
+					       pipe_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
```

Comment:
if pipe_id < 0, or comp_id < 0, it is an invalid user input isnt it? should we throw an error then?

Reply:
pipe_id < 0, comp_id < 0 and uuid_id == 0 are valid inputs and means "don't care". It allows to select every component in single message `log_level 0 -1 -1`, or each volume component: `log_level volume_id -1 -1`, or each component on selected pipeline: `log_level 0 pipe_id -1`. Unused values are not part of IPC message, it's why I use `TRACE_FILTER_MAX_IPC_ELEMENTS_PER_ENTRY` instead `TRACE_FILTER_IPC_ELEMENTS_PER_ENTRY`.